File: sdk/runanywhere-swift/Sources/RunAnywhere/Capabilities/Voice/Services/VoiceCapabilityService.swift
import Foundation
import os

/// Main capability coordinator for voice processing using the new component system
public class VoiceCapabilityService {
    private let logger = SDKLogger(category: "VoiceCapabilityService")

    // Services
    private let sessionManager: VoiceSessionManager
    private var analyticsService: VoiceAnalyticsService?
    private var sttAnalyticsService: STTAnalyticsService?

    // State
    private var isInitialized = false

    // Active voice agents
    private var activeAgents: [UUID: VoiceAgentComponent] = [:]
    private let agentQueue = DispatchQueue(label: "com.runanywhere.voiceagents", attributes: .concurrent)

    public init() {
        self.sessionManager = VoiceSessionManager()
    }

    /// Initialize the voice capability
    public func initialize() async throws {
        guard !isInitialized else {
            logger.debug("Voice capability already initialized")
            return
        }

        logger.info("Initializing voice capability")

        // Get analytics services from container
        analyticsService = await ServiceContainer.shared.voiceAnalytics
        sttAnalyticsService = await ServiceContainer.shared.sttAnalytics

        // Initialize sub-services
        await sessionManager.initialize()

        isInitialized = true
        logger.info("Voice capability initialized successfully")
    }

    /// Create a voice agent with the given parameters
    /// - Parameters:
    ///   - vadParams: VAD parameters (optional)
    ///   - sttParams: STT parameters (optional)
    ///   - llmParams: LLM parameters (optional)
    ///   - ttsParams: TTS parameters (optional)
    /// - Returns: Configured voice agent component
    @MainActor
    public func createVoiceAgent(
        vadParams: VADInitParameters? = nil,
        sttParams: STTInitParameters? = nil,
        llmParams: LLMInitParameters? = nil,
        ttsParams: TTSInitParameters? = nil
    ) async throws -> VoiceAgentComponent {
        logger.debug("Creating voice agent with custom parameters")

        // Create agent parameters (all components are required, use defaults if not provided)
        let agentParams = VoiceAgentInitParameters(
            vadParameters: vadParams ?? VADInitParameters(),
            sttParameters: sttParams ?? STTInitParameters(),
            llmParameters: llmParams ?? LLMInitParameters(),
            ttsParameters: ttsParams ?? TTSInitParameters()
        )

        // Create and initialize agent
        let agent = VoiceAgentComponent(
            parameters: agentParams,
            serviceContainer: ServiceContainer.shared
        )

        try await agent.initialize(with: agentParams)

        // Track agent
        let agentId = UUID()
        agentQueue.async(flags: .barrier) { [weak self] in
            self?.activeAgents[agentId] = agent
        }

        // Track analytics
        Task {
            await analyticsService?.trackPipelineCreation(
                stages: [
                    vadParams != nil ? "vad" : nil,
                    sttParams != nil ? "stt" : nil,
                    llmParams != nil ? "llm" : nil,
                    ttsParams != nil ? "tts" : nil
                ].compactMap { $0 }
            )
        }

        return agent
    }

    /// Create a full voice pipeline with all components
    public func createFullPipeline(
        sttModelId: String? = nil,
        llmModelId: String? = nil
    ) async throws -> VoiceAgentComponent {
        return try await createVoiceAgent(
            vadParams: VADInitParameters(),
            sttParams: STTInitParameters(modelId: sttModelId),
            llmParams: LLMInitParameters(modelId: llmModelId),
            ttsParams: TTSInitParameters()
        )
    }

    /// Process voice with a custom pipeline configuration
    public func processVoice(
        audioStream: AsyncStream<VoiceAudioChunk>,
        vadParams: VADInitParameters? = nil,
        sttParams: STTInitParameters? = nil,
        llmParams: LLMInitParameters? = nil,
        ttsParams: TTSInitParameters? = nil
    ) -> AsyncThrowingStream<VoiceAgentEvent, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    // Create agent
                    let agent = try await createVoiceAgent(
                        vadParams: vadParams,
                        sttParams: sttParams,
                        llmParams: llmParams,
                        ttsParams: ttsParams
                    )

                    // Convert audio chunks to Data stream
                    let dataStream = AsyncStream<Data> { dataContinuation in
                        Task {
                            for await chunk in audioStream {
                                dataContinuation.yield(chunk.data)
                            }
                            dataContinuation.finish()
                        }
                    }

                    // Process through agent
                    let eventStream = agent.processStream(dataStream)

                    for try await event in eventStream {
                        continuation.yield(event)
                    }

                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    /// Find voice service for a specific model
    @MainActor
    public func findVoiceService(for modelId: String?) async -> STTService? {
        // Check if any active agent has STT with the specified model
        let agents = agentQueue.sync { Array(activeAgents.values) }
        for agent in agents {
            if let stt = agent.sttComponent?.getService() {
                return stt
            }
        }
        return nil
    }

    /// Find LLM service for a specific model
    @MainActor
    public func findLLMService(for modelId: String?) async -> LLMService? {
        // Check if any active agent has LLM with the specified model
        let agents = agentQueue.sync { Array(activeAgents.values) }
        for agent in agents {
            if let llm = agent.llmComponent?.getService() {
                return llm
            }
        }
        return nil
    }

    /// Find TTS service
    @MainActor
    public func findTTSService() async -> TextToSpeechService? {
        // Check if any active agent has TTS
        let agents = agentQueue.sync { Array(activeAgents.values) }
        for agent in agents {
            if let tts = agent.ttsComponent?.getService() {
                return tts
            }
        }
        return nil
    }

    /// Clean up all active agents
    public func cleanup() async throws {
        let agents = agentQueue.sync { Array(activeAgents.values) }

        for agent in agents {
            try await agent.cleanup()
        }

        agentQueue.async(flags: .barrier) { [weak self] in
            self?.activeAgents.removeAll()
        }
    }
}

// MARK: - Backward Compatibility

extension VoiceCapabilityService {
    /// Legacy method for backward compatibility
    public func createPipeline(
        vadParams: VADInitParameters? = nil,
        sttParams: STTInitParameters? = nil,
        llmParams: LLMInitParameters? = nil,
        ttsParams: TTSInitParameters? = nil
    ) -> VoicePipelineManager {
        // Create a legacy wrapper around the new voice agent
        Task {
            let agent = try? await createVoiceAgent(
                vadParams: vadParams,
                sttParams: sttParams,
                llmParams: llmParams,
                ttsParams: ttsParams
            )

            // Store reference for later use
            if let agent = agent {
                logger.debug("Created legacy pipeline wrapper for voice agent")
            }
        }

        // Create a ModularPipelineConfig from the provided parameters
        let config = ModularPipelineConfig(
            components: [.vad, .stt, .llm, .tts],
            vad: vadParams,
            stt: sttParams,
            llm: llmParams,
            tts: ttsParams
        )

        // Return a pipeline manager with the config
        return VoicePipelineManager(config: config)
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/capabilities/voice/voice_capability_service.dart
/// Voice Capability Service
///
/// Main capability coordinator for voice processing using the component system.
/// Matches iOS VoiceCapabilityService.swift from Capabilities/Voice/Services/
library voice_capability_service;

import 'dart:async';
import 'dart:typed_data';

import 'package:runanywhere/capabilities/voice/services/voice_session_manager.dart';
import 'package:runanywhere/core/module_registry.dart'
    show ModuleRegistry, STTService, LLMService;
import 'package:runanywhere/features/llm/llm_capability.dart'
    show LLMConfiguration;
import 'package:runanywhere/features/stt/stt_capability.dart'
    show STTConfiguration;
import 'package:runanywhere/features/tts/models/tts_configuration.dart'
    show TTSConfiguration;
import 'package:runanywhere/features/tts/protocol/tts_service.dart'
    show TTSService;
import 'package:runanywhere/features/vad/vad_configuration.dart'
    show VADConfiguration;
import 'package:runanywhere/features/voice_agent/voice_agent_capability.dart';
import 'package:runanywhere/foundation/dependency_injection/service_container.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';

/// Voice audio chunk for streaming
class VoiceAudioChunk {
  final Uint8List data;
  final Duration timestamp;

  VoiceAudioChunk({required this.data, required this.timestamp});
}

/// Main capability coordinator for voice processing using the component system
/// Matches iOS VoiceCapabilityService class from VoiceCapabilityService.swift
class VoiceCapabilityService {
  final SDKLogger _logger = SDKLogger(category: 'VoiceCapabilityService');

  // Services
  final VoiceSessionManager _sessionManager;

  // State
  bool _isInitialized = false;

  // Active voice agents
  final Map<String, VoiceAgentCapability> _activeAgents = {};

  VoiceCapabilityService() : _sessionManager = VoiceSessionManager();

  /// Initialize the voice capability
  Future<void> initialize() async {
    if (_isInitialized) {
      _logger.debug('Voice capability already initialized');
      return;
    }

    _logger.info('Initializing voice capability');

    // Initialize sub-services
    await _sessionManager.initialize();

    _isInitialized = true;
    _logger.info('Voice capability initialized successfully');
  }

  /// Create a voice agent with the given parameters
  Future<VoiceAgentCapability> createVoiceAgent({
    VADConfiguration? vadParams,
    STTConfiguration? sttParams,
    LLMConfiguration? llmParams,
    TTSConfiguration? ttsParams,
  }) async {
    _logger.debug('Creating voice agent with custom parameters');

    // Create agent configuration (all components are required, use defaults if not provided)
    final agentConfig = VoiceAgentConfiguration(
      vadConfig: vadParams ?? const VADConfiguration(),
      sttConfig: sttParams ?? STTConfiguration(),
      llmConfig: llmParams ?? LLMConfiguration(),
      ttsConfig: ttsParams ?? const TTSConfiguration(),
    );

    // Create and initialize agent
    final agent = VoiceAgentCapability(
      configuration: agentConfig,
      serviceContainer: ServiceContainer.shared,
    );

    await agent.initialize();

    // Track agent
    final agentId = DateTime.now().millisecondsSinceEpoch.toString();
    _activeAgents[agentId] = agent;

    return agent;
  }

  /// Create a full voice pipeline with all components
  Future<VoiceAgentCapability> createFullPipeline({
    String? sttModelId,
    String? llmModelId,
  }) async {
    return createVoiceAgent(
      vadParams: const VADConfiguration(),
      sttParams: STTConfiguration(modelId: sttModelId),
      llmParams: LLMConfiguration(modelId: llmModelId),
      ttsParams: const TTSConfiguration(),
    );
  }

  /// Process voice with a custom pipeline configuration
  Stream<VoiceAgentEvent> processVoice({
    required Stream<VoiceAudioChunk> audioStream,
    VADConfiguration? vadParams,
    STTConfiguration? sttParams,
    LLMConfiguration? llmParams,
    TTSConfiguration? ttsParams,
  }) async* {
    try {
      // Create agent
      final agent = await createVoiceAgent(
        vadParams: vadParams,
        sttParams: sttParams,
        llmParams: llmParams,
        ttsParams: ttsParams,
      );

      // Convert audio chunks to Data stream
      final dataStream = audioStream.map((chunk) => chunk.data);

      // Process through agent
      await for (final event in agent.processStream(dataStream)) {
        yield event;
      }
    } catch (e) {
      yield VoiceAgentEvent.error(e);
    }
  }

  /// Find voice service for a specific model
  Future<STTService?> findVoiceService({String? modelId}) async {
    // Check if any active agent has STT with the specified model
    for (final agent in _activeAgents.values) {
      final stt = agent.sttCapability?.getService();
      if (stt != null) {
        return stt;
      }
    }

    // Fallback: try to create via provider
    final provider = ModuleRegistry.shared.sttProvider(modelId: modelId);
    if (provider == null) {
      _logger.warning('No STT provider available for model: $modelId');
      return null;
    }

    try {
      final config = STTConfiguration(
        modelId: modelId,
        language: 'en',
        sampleRate: 16000,
      );
      final service = await provider.createSTTService(config);
      return service;
    } catch (e) {
      _logger.error('Failed to create STT service: $e');
      return null;
    }
  }

  /// Find LLM service for a specific model
  Future<LLMService?> findLLMService({String? modelId}) async {
    // Check if any active agent has LLM with the specified model
    for (final agent in _activeAgents.values) {
      final llm = agent.llmCapability?.getService();
      if (llm != null) {
        return llm;
      }
    }
    return null;
  }

  /// Find TTS service
  Future<TTSService?> findTTSService() async {
    // Check if any active agent has TTS
    for (final agent in _activeAgents.values) {
      final tts = agent.ttsCapability?.getService();
      if (tts != null) {
        return tts;
      }
    }
    return null;
  }

  /// Get the session manager
  VoiceSessionManager get sessionManager => _sessionManager;

  /// Clean up all active agents
  Future<void> cleanup() async {
    for (final agent in _activeAgents.values) {
      await agent.cleanup();
    }
    _activeAgents.clear();
    _isInitialized = false;
  }
}

// MARK: - Backward Compatibility

extension VoiceCapabilityServiceBackwardCompat on VoiceCapabilityService {
  /// Create a voice pipeline using the new architecture
  Future<VoiceAgentCapability> createPipeline({
    VADConfiguration? vadParams,
    STTConfiguration? sttParams,
    LLMConfiguration? llmParams,
    TTSConfiguration? ttsParams,
  }) async {
    // Use the new VoiceAgent component which is the modern pipeline
    return createVoiceAgent(
      vadParams: vadParams,
      sttParams: sttParams,
      llmParams: llmParams,
      ttsParams: ttsParams,
    );
  }
}
